src/polyfem/solver/ALSolver.cpp

```diff
@@ -5,28 +5,24 @@
 namespace polyfem::solver
 {
 	ALSolver::ALSolver(
-		std::shared_ptr<NLSolver> nl_solver,
 		std::shared_ptr<BCLagrangianForm> lagr_form,
 		std::shared_ptr<BCPenaltyForm> pen_form,
 		const double initial_al_weight,
 		const double scaling,
 		const double max_al_weight,
 		const double eta_tol,
-		const int max_solver_iter,
 		const std::function<void(const Eigen::VectorXd &)> &update_barrier_stiffness)
-		: nl_solver(nl_solver),
-		  lagr_form(lagr_form),
+		: lagr_form(lagr_form),
 		  pen_form(pen_form),
 		  initial_al_weight(initial_al_weight),
 		  scaling(scaling),
 		  max_al_weight(max_al_weight),
 		  eta_tol(eta_tol),
-		  max_solver_iter(max_solver_iter),
 		  update_barrier_stiffness(update_barrier_stiffness)
 	{
 	}
 
-	void ALSolver::solve(NLProblem &nl_problem, Eigen::MatrixXd &sol)
+	void ALSolver::solve_al(std::shared_ptr<NLSolver> nl_solver,NLProblem &nl_problem, Eigen::MatrixXd &sol)
```

Comment:
why moving it from ctr to func?

Reply:
because we need a different solver for the last solve in ALSolver -- the newton reduced solve.